include/crab/domains/object_domain.hpp

```diff
@@ -0,0 +1,2378 @@
+    //        Perform assignment on base dom for res := rgn
+    //  or the number of reference > 1:
+    //  3.2 if object domain is Top, the overapproximate res value as top
+    //  3.3 if object domain exists, extract interval for rgn in object domain
+    //    assign the interval into base domain
+    //
+    //  post condition: the odi map not changed, the base domain is updated
+
+    ERROR_IF_NOT_REGION(rgn, __LINE__);
+    ERROR_IF_ARRAY_REGION(rgn, __LINE__);
+    ERROR_IF_NOT_REF(ref, __LINE__);
+    if ((rgn.get_type().is_bool_region() && !res.get_type().is_bool()) ||
+        (rgn.get_type().is_integer_region() && !res.get_type().is_integer()) ||
+        (rgn.get_type().is_real_region() && !res.get_type().is_real()) ||
+        (rgn.get_type().is_reference_region() &&
+         !res.get_type().is_reference())) {
+      CRAB_ERROR(domain_name(), "::ref_load: type of lhs ", res, " (",
+                 res.get_type(), ") is not compatible with region ", rgn, " (",
+                 rgn.get_type(), ")");
+    }
+    // At this point, the requirement for 1 is satisfied
+
+    if (is_bottom()) {
+      return;
+    }
+
+    // for now skip analysis for unknown region
+    if (is_unknown_region(rgn)) {
+      return;
+    }
+
+    if (is_null_ref(ref).is_true()) {
+      CRAB_LOG("object", CRAB_WARN(domain_name(), "::ref_load: reference ", ref,
+                                   " is null."););
+      // set_to_bottom();
+      crab::CrabStats::count(domain_name() +
+                             ".count.ref_load.skipped.null_reference");
+      m_base_dom -= res;
+      return;
+    }
+    // At this point, the requirement for 2 is satisfied
+
+    if (auto id_opt = get_obj_id(rgn)) {
+      // the object id must exist for each region if the
+      // region belongs to an abstract object.
+
+      // retrieve an abstract object info
+      auto obj_info_ref = m_obj_info_env.find(*id_opt);
+      assert(obj_info_ref); // The object info must exsits
+
+      const small_range &num_refs = (*obj_info_ref).refcount_val();
+
+      // In crab IR, the number of references cannot be zero
+      //  if ref_load access a not null reference.
+      // So zero case should not exist
+      assert(!num_refs.is_zero());
+
+      if (num_refs.is_one()) { // singleton object
+        m_base_dom.assign(res, rgn);
+        // the requirment 3.1 is satisfied
+      } else { // num_refs > 1, use odi map
+        // read from odi map
+        const odi_domain_product_t *prod_ref = m_odi_map.find((*id_opt));
+        if (prod_ref != nullptr) {
+          term::term_operator_t symb =
+              invalidate_cache_if_miss((*id_opt), ref, rgn);
+          if (res.get_type().is_reference()) {
+            // res is a reference, assign a fresh symbol into address dom
+            // create a fresh symbol to represent its base address
+            m_addrs_dom.set(get_or_insert_base_addr(res),
```

Comment:
Why new symbol? It should be the same symbol as the base address of what is currently loaded in the cache